File: benchmarks-website/server/src/api/descriptions.rs
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: Copyright the Vortex contributors

//! Editorial group descriptions, ported from v2.
//!
//! These strings are the source of truth for the hover tooltip rendered on
//! every `<details>` group title on the landing page and on the
//! `/group/{slug}` permalink. They are deliberately editorial and
//! hand-maintained — derived from the group *name*, not from the database —
//! so that adding a new group's blurb is a one-line edit here rather than a
//! schema or ingest change.
//!
//! Source of truth in v2 (kept verbatim where applicable):
//! - `benchmarks-website/src/utils.js` — `getBenchmarkDescription`
//! - `benchmarks-website/src/config.js` — `BENCHMARK_DESCRIPTIONS`,
//!   `QUERY_SUITES.description`
//!
//! TPC-H / TPC-DS fan out by storage and scale factor; the description is
//! synthesised from the parsed name so we don't have to hand-maintain one
//! entry per `(storage, sf)` pair.

/// Look up a short editorial description for a group display name. Returns
/// `None` when the group has no canonical description (e.g. vector-search
/// groups) — callers render the title without a tooltip in that case.
pub fn group_description(name: &str) -> Option<String> {
    if let Some(d) = tpc_description(name) {
        return Some(d);
    }
    static_description(name).map(str::to_string)
}

/// Hard-coded, name-keyed descriptions for the non-fan-out groups. These
/// match v2 verbatim where v2 had a description; new strings here should
/// match the wording style v2 set.
fn static_description(name: &str) -> Option<&'static str> {
    match name {
        "Random Access" => {
            Some("Tests performance of selecting arbitrary row indices from a file on NVMe storage")
        }
        "Compression" => Some(
            "Measures encoding and decoding throughput (MB/s) for Vortex files and Parquet \
             files (with zstd page compression)",
        ),
        "Compression Size" => Some(
            "Compares compressed file sizes and compression ratios across different encoding \
             strategies",
        ),
        "Clickbench" => Some(
            "ClickHouse's analytical benchmark suite testing real-world query patterns on web \
             analytics data",
        ),
        "Statistical and Population Genetics" => {
            Some("A suite of Statistical and Population genetics queries using the gnomAD dataset")
        }
        "PolarSignals Profiling" => Some(
            "Profiling data benchmark modeled on PolarSignals/Parca, exercising scan-layer \
             performance with projection and filter pushdown on deeply nested schemas",
        ),
        _ => None,
    }
}

/// Derive a description for `TPC-H (NVMe|S3) (SF=N)` and `TPC-DS (NVMe) (SF=N)`
/// group names. The shape is fixed because [`crate::api::groups::group_name_query`]
/// emits exactly this format for tpch/tpcds. Returns `None` for any name that
/// does not start with `TPC-H ` or `TPC-DS `.
fn tpc_description(name: &str) -> Option<String> {
    let parts = if let Some(rest) = name.strip_prefix("TPC-H ") {
        Some(("TPC-H", rest))
    } else {
        name.strip_prefix("TPC-DS ").map(|rest| ("TPC-DS", rest))
    };
    let (suite, rest) = parts?;
    let storage = if rest.starts_with("(NVMe)") {
        "nvme"
    } else if rest.starts_with("(S3)") {
        "s3"
    } else {
        return None;
    };
    let sf = parse_sf(rest)?;
    Some(format_tpc(suite, storage, &sf))
}

/// Pull `SF=N` (digits only) out of strings like `(NVMe) (SF=10)`. Returns
/// `None` if no `SF=` substring or the digits don't parse.
fn parse_sf(s: &str) -> Option<String> {
    let after = s.split_once("SF=")?.1;
    let digits: String = after.chars().take_while(char::is_ascii_digit).collect();
    if digits.is_empty() {
        None
    } else {
        Some(digits)
    }
}

/// Render the v2-compatible TPC blurb. Storage label comes from the parsed
/// group name; scale-bytes annotation only renders for TPC-H (TPC-DS in v2
/// did not annotate scale bytes).
fn format_tpc(suite: &str, storage: &str, sf: &str) -> String {
    let storage_phrase = match storage {
        "nvme" => "on local NVMe storage",
        "s3" => "against S3 storage",
        _ => "on local NVMe storage",
    };
    let bytes = match sf {
        "1" => Some("1GB"),
        "10" => Some("10GB"),
        "100" => Some("100GB"),
        "1000" => Some("1TB"),
        _ => None,
    };
    match (suite, bytes) {
        ("TPC-H", Some(b)) => {
            format!("TPC-H benchmark queries {storage_phrase} at SF={sf} (~{b} of data)",)
        }
        ("TPC-H", None) => format!("TPC-H benchmark queries {storage_phrase} at SF={sf}"),
        ("TPC-DS", _) => format!("TPC-DS benchmark queries {storage_phrase} at SF={sf}"),
        _ => format!("{suite} benchmark queries {storage_phrase} at SF={sf}"),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn static_descriptions_match_v2() {
        assert_eq!(
            group_description("Random Access").as_deref(),
            Some(
                "Tests performance of selecting arbitrary row indices from a file on NVMe storage"
            ),
        );
        assert_eq!(
            group_description("Compression").as_deref(),
            Some(
                "Measures encoding and decoding throughput (MB/s) for Vortex files and Parquet \
                 files (with zstd page compression)",
            ),
        );
        assert_eq!(
            group_description("Compression Size").as_deref(),
            Some(
                "Compares compressed file sizes and compression ratios across different encoding \
                 strategies",
            ),
        );
        assert_eq!(
            group_description("Clickbench").as_deref(),
            Some(
                "ClickHouse's analytical benchmark suite testing real-world query patterns on \
                 web analytics data",
            ),
        );
        assert_eq!(
            group_description("Statistical and Population Genetics").as_deref(),
            Some("A suite of Statistical and Population genetics queries using the gnomAD dataset",),
        );
        assert_eq!(
            group_description("PolarSignals Profiling").as_deref(),
            Some(
                "Profiling data benchmark modeled on PolarSignals/Parca, exercising scan-layer \
                 performance with projection and filter pushdown on deeply nested schemas",
            ),
        );
    }

    #[test]
    fn tpch_descriptions_carry_scale_bytes() {
        assert_eq!(
            group_description("TPC-H (NVMe) (SF=1)").as_deref(),
            Some("TPC-H benchmark queries on local NVMe storage at SF=1 (~1GB of data)"),
        );
        assert_eq!(
            group_description("TPC-H (S3) (SF=10)").as_deref(),
            Some("TPC-H benchmark queries against S3 storage at SF=10 (~10GB of data)"),
        );
        assert_eq!(
            group_description("TPC-H (NVMe) (SF=100)").as_deref(),
            Some("TPC-H benchmark queries on local NVMe storage at SF=100 (~100GB of data)"),
        );
        assert_eq!(
            group_description("TPC-H (S3) (SF=1000)").as_deref(),
            Some("TPC-H benchmark queries against S3 storage at SF=1000 (~1TB of data)"),
        );
    }

    #[test]
    fn tpcds_descriptions_omit_scale_bytes() {
        assert_eq!(
            group_description("TPC-DS (NVMe) (SF=1)").as_deref(),
            Some("TPC-DS benchmark queries on local NVMe storage at SF=1"),
        );
        assert_eq!(
            group_description("TPC-DS (NVMe) (SF=10)").as_deref(),
            Some("TPC-DS benchmark queries on local NVMe storage at SF=10"),
        );
    }

    #[test]
    fn unknown_groups_have_no_description() {
        assert_eq!(group_description("cohere-large-10m / partitioned"), None);
        assert_eq!(group_description("Made-up benchmark"), None);
    }

    #[test]
    fn malformed_tpc_names_fall_through() {
        // No `(NVMe)` / `(S3)` prefix → not matched.
        assert_eq!(group_description("TPC-H something else"), None);
        // SF= without digits → not matched.
        assert_eq!(group_description("TPC-H (NVMe) (SF=)"), None);
    }
}

File: benchmarks-website/web/lib/descriptions.ts
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: Copyright the Vortex contributors

/**
 * Editorial group descriptions, the TypeScript port of
 * `server/src/api/descriptions.rs` (originally ported from v2's
 * `src/config.js` `BENCHMARK_DESCRIPTIONS` + `src/utils.js`
 * `getBenchmarkDescription`).
 *
 * These strings are the source of truth for the hover tooltip rendered on each
 * group title. They are deliberately editorial and hand-maintained, derived
 * from the group *name* rather than from the database, so adding a new group's
 * blurb is a one-line edit here rather than a schema or ingest change.
 *
 * `TPC-H` / `TPC-DS` group names fan out by storage and scale factor, so their
 * descriptions are synthesised from the parsed name rather than hand-listed per
 * `(storage, sf)` pair.
 */

/**
 * Look up a short editorial description for a group display name. Returns
 * `null` when the group has no canonical description (e.g. vector-search
 * groups); callers render the title without a tooltip in that case.
 *
 * Mirrors the Rust `group_description`: the synthesised `TPC-*` blurb takes
 * precedence, then the hard-coded name-keyed table.
 */
export function groupDescription(name: string): string | null {
  const tpc = tpcDescription(name);
  if (tpc !== null) {
    return tpc;
  }
  return staticDescription(name);
}

/**
 * Hard-coded, name-keyed descriptions for the non-fan-out groups. These match
 * v2 verbatim where v2 had a description.
 */
function staticDescription(name: string): string | null {
  switch (name) {
    case 'Random Access':
      return (
        'Point lookups — selecting specific rows by position from an NVMe file. What feature ' +
        'stores, vector retrieval, and per-record serving actually do.'
      );
    case 'Compression':
      return (
        'Encode and decode throughput (MB/s) for Vortex vs Parquet (zstd page compression). ' +
        'Encode gates ingestion; decode gates every scan after.'
      );
    case 'Compression Size':
      return (
        'Compressed file size per format across a fixed set of datasets. A faster format that ' +
        'bloats on disk just trades one bill for another.'
      );
    case 'Clickbench':
      return (
        "ClickHouse's 43-query analytical suite over real web-analytics data — the field's " +
        'standard test for single-table scans, filters, and aggregations.'
      );
    case 'Statistical and Population Genetics':
      return (
        'Population-genetics queries over the gnomAD dataset — DuckDB-only, exercising the ' +
        'deeply-nested array operations real genomics pipelines run on.'
      );
    case 'PolarSignals Profiling':
      return (
        'Scan-layer benchmark modeled on PolarSignals/Parca: projection and filter pushdown ' +
        'over deeply-nested profile schemas — the shape continuous-profiling backends actually read.'
      );
    default:
      return null;
  }
}

/**
 * Suite-level TPC blurb (no storage / scale-factor dimensions) for the clustered
 * Historic-page section, where those become toggle buttons rather than part of
 * the heading. Returns `null` for non-TPC names.
 */
export function tpcSuiteDescription(suite: string): string | null {
  switch (suite) {
    case 'TPC-H':
      return 'TPC-H — 22 analytical queries against the canonical OLAP star schema.';
    case 'TPC-DS':
      return (
        'TPC-DS — the broader 99-query analytical suite, with larger schemas and skewed ' +
        'distributions.'
      );
    default:
      return null;
  }
}

/**
 * Derive a description for `TPC-H (NVMe|S3) (SF=N)` and `TPC-DS (NVMe) (SF=N)`
 * group names. The shape is fixed because the query-group name builder emits
 * exactly this format for tpch/tpcds. Returns `null` for any name that does not
 * start with `TPC-H ` or `TPC-DS `.
 */
function tpcDescription(name: string): string | null {
  let suite: 'TPC-H' | 'TPC-DS';
  let rest: string;
  if (name.startsWith('TPC-H ')) {
    suite = 'TPC-H';
    rest = name.slice('TPC-H '.length);
  } else if (name.startsWith('TPC-DS ')) {
    suite = 'TPC-DS';
    rest = name.slice('TPC-DS '.length);
  } else {
    return null;
  }
  let storage: 'nvme' | 's3';
  if (rest.startsWith('(NVMe)')) {
    storage = 'nvme';
  } else if (rest.startsWith('(S3)')) {
    storage = 's3';
  } else {
    return null;
  }
  const sf = parseSf(rest);
  if (sf === null) {
    return null;
  }
  return formatTpc(suite, storage, sf);
}

/**
 * Pull `SF=N` (leading digits only) out of strings like `(NVMe) (SF=10)`.
 * Returns `null` if there is no `SF=` substring or the digits don't parse.
 */
function parseSf(s: string): string | null {
  const idx = s.indexOf('SF=');
  if (idx === -1) {
    return null;
  }
  const after = s.slice(idx + 'SF='.length);
  let digits = '';
  for (const ch of after) {
    if (ch >= '0' && ch <= '9') {
      digits += ch;
    } else {
      break;
    }
  }
  return digits.length === 0 ? null : digits;
}

/**
 * Render the v2-compatible TPC blurb. The storage phrase comes from the parsed
 * group name; the scale-bytes annotation only renders for TPC-H (v2's TPC-DS
 * descriptions did not annotate scale bytes).
 */
function formatTpc(suite: 'TPC-H' | 'TPC-DS', storage: 'nvme' | 's3', sf: string): string {
  const storagePhrase = storage === 's3' ? 'against S3' : 'on local NVMe';
  const bytes = scaleBytes(sf);
  if (suite === 'TPC-H') {
    const schema = 'TPC-H — 22 analytical queries against the canonical OLAP star schema —';
    if (bytes !== null) {
      return `${schema} at scale factor ${sf} (~${bytes}), ${storagePhrase}.`;
    }
    return `${schema} at scale factor ${sf}, ${storagePhrase}.`;
  }
  return (
    'TPC-DS — the broader 99-query analytical suite (larger schemas, skewed distributions) — ' +
    `at scale factor ${sf}, ${storagePhrase}.`
  );
}

/** The human-readable dataset size for the known TPC scale factors. */
function scaleBytes(sf: string): string | null {
  switch (sf) {
    case '1':
      return '1 GB';
    case '10':
      return '10 GB';
    case '100':
      return '100 GB';
    case '1000':
      return '1 TB';
    default:
      return null;
  }
}
